articles/node_lifecycle.md

```diff
@@ -0,0 +1,176 @@
+
+Any service requests to a node in the inactive state will not be answered (to the caller, they will fail immediately).
+
+Entering the `inactive` state from the `created` state requires running the node's `onInitialised` method. The node uses this to set up any resources it must hold throughout its life (irrespective of if it is active or inactive). As examples, such resources may include topic publications and subscriptions, memory that is held continuously, and initialising configuration parameters.
+
+Entering the `inactive` state from the `active` state requires running the node's `onDeactivated` method. See the `active` state description for more information.
+
+Entering the `inactive` state from the `fatal error` state requires running the node's `onReset` method. See the `fatal error` state description for more information.
+
+If an error occurs while in this state that cannot be handled by the node, then the transition to the `fatal error` state is taken.
+
+The node may also exit this state by being destroyed, in which case the `destroy` transition is taken.
+
+Implementation note: Causing the transition to the destroyed state can often be mapped to deletion of the node instance.
+
+### Active state
+
+This is the main state of the node's life cycle. While in this state, the node performs any processing, responds to service requests, reads and processes data, produces output, etc.
+
+Entering the `active` state requires running the node's `onActivated` method. This is used to prepare the node for execution. This may include acquiring resources that are only held while the node is actually active, such as access to hardware. Ideally, no preparation that requires significant time (such as lengthy hardware initialisation) should be performed in this callback.
+
+To shut down a node, it must first be deactivated. This will cause the `onDeactivate` method to be called, allowing the node to clean up any resources it only holds while active.
+
+If an error that cannot be handled by the node/system occurs in this state, the transition to the `fatal error` state is taken.
+
+### Fatal Error state
+
+Known errors that can occur in a node should be considered designed behaviour of the system and so should be handled by the node where the error occurred and, if necessary, the surrounding nodes. The reason for this is that the system knows best how to handle errors that occur in its behaviour.
+
+Any fatal errors (those which the node/system was not expecting and does not know how to handle) will cause a transition to the `fatal error` state. In this state, because the node does not know what to do in the presence of the error, it is not provided any execution time. Arriving data is not processed, service requests are not responded to, and no output is produced.
+
+On entering the `fatal error` state, the node's `onError` method must be executed. This provides the node with a chance to clean up resources it was using.
```

Comment:
>  For example, there could be an error which the node simply does not handle, which requires the supervisor to destroy and recreate the node (or do some other behavior), but does not necessarily mean the supervisor or other nodes in the same process must terminate.

I start to sense that we are still talking about two types of fatal errors ...

How it works in RTT is: if a shutdown transition (Active->Inactive, Inactive->Created, XXX->Exception) raises, the node is transitioned to Fatal. The rationale is that by interrupting the shutdown process, the component ends up in an inconsistent state.

What you seem to talk about is having a state that is declared by the component as "non recoverable".

I do agree that there is theoretically room for both. However, I really never had the need for the second use case, since doing a full active->initialized->created->initialized->active cycle is meant to do a full reset of the component, the only purely fatal error is the one where the transitions themselves have uncaught exceptions (which transition to the RTT-style Fatal)

Anyways, if what you mean is that you want a possible transition from Fatal to Destroyed, I personally don't see a downside to that OK with that as long as the supervisor can decide what it should do.

NB: bringing down a process it not necessarily "fatal" to the supervisor. At least, it's not on my systems.